src/tsdb.c

```diff
@@ -10,83 +10,104 @@
     newSeries->maxSamplesPerChunk = maxSamplesPerChunk;
-    newSeries->firstChunk = NewChunk(newSeries->maxSamplesPerChunk);
-    newSeries->lastChunk = newSeries->firstChunk;
-    newSeries->chunkCount = 1;
     newSeries->retentionSecs = retentionSecs;
     newSeries->rules = NULL;
     newSeries->lastTimestamp = 0;
     newSeries->lastValue = 0;
-
+    Chunk* newChunk = NewChunk(newSeries->maxSamplesPerChunk);
+    RedisModule_DictSetC(newSeries->chunks, (void*)&newSeries->lastTimestamp, sizeof(newSeries->lastTimestamp),
+                        (void*)newChunk);
     return newSeries;
 }
 
 void SeriesTrim(Series * series) {
     if (series->retentionSecs == 0) {
         return;
     }
-    Chunk *currentChunk = series->firstChunk;
+
+    // start iterator from smallest key
+    RedisModuleDictIter *iter = RedisModule_DictIteratorStartC(series->chunks, "^", NULL, 0);
+    Chunk *currentChunk;
+    void *currentKey;
     timestamp_t minTimestamp = time(NULL) - series->retentionSecs;
-    while (currentChunk != NULL)
+    while ((currentKey=RedisModule_DictNextC(iter, NULL, (void*)&currentChunk)))
     {
         if (ChunkGetLastTimestamp(currentChunk) < minTimestamp)
         {
-            Chunk *nextChunk = currentChunk->nextChunk;
-            if (nextChunk != NULL) {
-                series->firstChunk = nextChunk;    
-            } else {
-                series->firstChunk = NewChunk(series->maxSamplesPerChunk);
-            }
-            
-            series->chunkCount--;
+            RedisModule_DictDelC(series->chunks, currentKey, sizeof(timestamp_t), NULL);
+            // reseek iterator since we modified the dict, go to first element that is bigger than current key
+            RedisModule_DictIteratorReseekC(iter, ">", currentKey, sizeof(timestamp_t));
             FreeChunk(currentChunk);
-            currentChunk = nextChunk;
         } else {
             break;
         }
     }
+    RedisModule_DictIteratorStop(iter);
 }
 
 void FreeSeries(void *value) {
     Series *currentSeries = (Series *) value;
-    Chunk *currentChunk = currentSeries->firstChunk;
-    while (currentChunk != NULL)
+    RedisModuleDictIter *iter = RedisModule_DictIteratorStartC(currentSeries->chunks, "^", NULL, 0);
+    Chunk *currentChunk;
+    void *currentKey;
+    while ((currentKey=RedisModule_DictNextC(iter, NULL, (void*)&currentChunk)))
     {
-        Chunk *nextChunk = currentChunk->nextChunk;
         FreeChunk(currentChunk);
-        currentChunk = nextChunk;
+        RedisModule_DictIteratorReseekC(iter, ">", currentKey, sizeof(currentKey));
     }
+    RedisModule_DictIteratorStop(iter);
+    RedisModule_FreeDict(NULL, currentSeries->chunks);
 }
 
 size_t SeriesMemUsage(const void *value) {
     Series *series = (Series *)value;
-    return sizeof(series) + sizeof(Chunk) * series->chunkCount;
+    return sizeof(series) + sizeof(Chunk) * RedisModule_DictSize(series->chunks);
+}
+
+Chunk *SeriesGetLastChunk(Series *series)
```

Comment:
@yaeltzirulnikov whats the complexity for this function?
this has influence on ts.add complexity, we might want to "cache" the pointer to the last pointer.